src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -727,11 +744,159 @@ BoundExpression convertArrayElement(BoundExpression element, TypeSymbol elementT
+                    node,
+                    CollectionLiteralTypeKind.CollectionBuilder,
+                    GetWellKnownType(WellKnownType.System_Collections_Generic_List_T, diagnostics, syntax).Construct(elementType),
+                    wasCompilerGenerated: wasCompilerGenerated,
+                    diagnostics);
+                return result.Update(result.CollectionTypeKind, result.Placeholder, result.CollectionCreation, constructMethod, result.Elements, targetType);
+            }
+
+            var implicitReceiver = new BoundObjectOrCollectionValuePlaceholder(syntax, isNewInstance: true, spanType) { WasCompilerGenerated = true };
+            var builder = ArrayBuilder<BoundExpression>.GetInstance(elements.Length);
+            foreach (var element in elements)
+            {
+                builder.Add(ConvertCollectionLiteralArrayElement(element, elementType, diagnostics));
+            }
+            return new BoundCollectionLiteralExpression(
+                syntax,
+                CollectionLiteralTypeKind.CollectionBuilder,
+                implicitReceiver,
+                collectionCreation: null,
+                collectionBuilderMethod: constructMethod,
+                builder.ToImmutableAndFree(),
+                targetType)
+            { WasCompilerGenerated = wasCompilerGenerated };
+        }
+
+        private MethodSymbol? GetCollectionBuilderMethod(
+            ExpressionSyntax syntax,
+            NamedTypeSymbol targetType,
+            TypeSymbol elementTypeOriginalDefinition,
+            TypeSymbol? builderType,
+            string? methodName,
+            BindingDiagnosticBag diagnostics)
+        {
+            if (!(builderType is NamedTypeSymbol { TypeKind: TypeKind.Class or TypeKind.Struct, IsGenericType: false }))
+            {
+                return null;
+            }
+
+            if (string.IsNullOrEmpty(methodName))
+            {
+                return null;
+            }
+
+            foreach (var candidate in builderType.GetMembers(methodName))
+            {
+                var method = candidate as MethodSymbol;
+                if (method is null || !method.IsStatic)
+                {
+                    continue;
+                }
+
+                var useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
+                if (!IsAccessible(method, ref useSiteInfo))
+                {
+                    continue;
+                }
+
+                var builder = ArrayBuilder<TypeWithAnnotations>.GetInstance();
+                targetType.GetAllTypeArgumentsNoUseSiteDiagnostics(builder);
+                var allTypeArguments = builder.ToImmutableAndFree();
+
+                if (method.Arity != allTypeArguments.Length)
+                {
+                    continue;
+                }
+
+                if (!(method.Parameters is [var parameter] && isReadOnlySpanType(Compilation, parameter.Type)))
```

Comment:
Nit: I personally would find this easier to read as `is not || !isReadOnlySpanType`, but up to you if you think this is better.

Reply:
agree with fred on this.  mixing `!` and patterns is always confusing to me (since we have `not`).